idea/src/org/jetbrains/kotlin/idea/inspections/SuspiciousEqualsCombination.kt

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright 2010-2017 JetBrains s.r.o.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.jetbrains.kotlin.idea.inspections
+
+import com.intellij.codeInspection.CleanupLocalInspectionTool
+import com.intellij.codeInspection.LocalInspectionToolSession
+import com.intellij.codeInspection.ProblemHighlightType
+import com.intellij.codeInspection.ProblemsHolder
+import org.jetbrains.kotlin.lexer.KtTokens
+import org.jetbrains.kotlin.psi.*
+import java.util.*
+
+class SuspiciousEqualsCombination : AbstractKotlinInspection(), CleanupLocalInspectionTool {
+    override fun buildVisitor(holder: ProblemsHolder, isOnTheFly: Boolean, session: LocalInspectionToolSession) =
+            object : KtVisitorVoid() {
+                override fun visitBinaryExpression(expression: KtBinaryExpression) {
+                    if (expression.parent is KtBinaryExpression) return
+                    val (first, second) = expression.parseBinary()
+                    val eqeq = first.map { it.text }
+                    val eqeqeq = second.map { it.text }
+                    if (eqeq.fold(false) { acc, it -> acc || eqeqeq.contains(it) }) {
+                        holder.registerProblem(expression, "Suspicious combination of == and ===",
+                                               ProblemHighlightType.GENERIC_ERROR_OR_WARNING)
+                    }
+                }
+            }
+
+    private fun KtBinaryExpression.parseBinary(pair: Pair<MutableList<KtExpression>, MutableList<KtExpression>>
```

Comment:
I think the intent of this code will be much clearer if you used an explicit class instead of a Pair:

```
data class ComparisonOperands(val eqEqOperands: MutableList<KtExpression>, val eqEqEqOperands: MutableList<KtExpression>)